Javascript: Use function literals in arguments to detect function composition

brainkim opened the pull request:
<!-- Please provide a brief summary of your changes: -->
This PR removes the hard-coded strings used to detect function composition in favor of a heuristic which counts the number of function literals passed as arguments.

In my opinion, the counter-examples are all places where it would be unreasonable to expect line-breaks (e.g. redux connect functions) or where line-breaks should occur anyways (the most common example is calling `Promise.prototype.then` with two short functions).

<!-- Please ensure you’ve done all of these things (if applicable). -->
<!-- You can replace the `[ ]` with `[x]` to mark each task as done. -->

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to the `CHANGELOG.unreleased.md` file following the template.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

Related issues:
- Long curried function calls #1420
- RxJS pipe chaining get formatted on a single line #4172
- Add heuristic to format functional composition more nicely #4431
- Functional composition heuristics false positives #4751
- Improve functional composition heuristics #4758
- Codes format unexpected with function "pipe" #4858
- Composed functions not put on own line anymore after 1.14.0 update #4925
- Strange behavior with word `connect` #4935
- Functional Composition Heuristic Is Too Broad #5060
- fix(javascript): skip .connect() method when composing fun #5739
- Prettier formats a sensible one-liner into seven lines #5769
- Inconsitent formatting for one specific line (always the same line) #5898
- Ugly object destructing and function #5928
- Calls to functions or methods named “compose” are always split into multiple lines. #5969
- Add a way to disable functional composition rule #6030

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

fixes #4751 fixes #4858 fixes #5060 fixes #5769 fixes #5928 fixes #5969

brainkim commented:
The motivating intuition behind this heuristic is that inlined function composition only becomes unreadable when there are multiple anonymous functions within the same call.

Consider the following example from one of the earliest issues (https://github.com/prettier/prettier/issues/4172):

```js
import { range } from 'rxjs/observable/range';
import { map, filter, scan } from 'rxjs/operators';

const source$ = range(0, 10);

source$.pipe(
  filter(x => x % 2 === 0),
  map(x => x + x),
  scan((acc, x) => acc + x, 0)
)
.subscribe(x => console.log(x))
```

If we named the functions, it wouldn’t look so ugly for the function composition to happen on the same line:
```js
import { range } from 'rxjs/observable/range';
import { map, filter, scan } from 'rxjs/operators';

const source$ = range(0, 10);
const even = x => x % 2 === 0;
const double = x => x + x;
const add = (acc, x) => acc + x, 0;
// easy to understand, isn’t split up using new heuristic
source$
  .pipe(filter(even), map(double), scan(add))
  .subscribe(x => console.log(x))
```

Rather than forcing authors to name their functions we just check for function literals.

brainkim commented:
@duailibe Who do I ping to get this PR considered?
@suchipi original PR author?

evilebottnawi commented:
/cc @prettier/core

evilebottnawi commented on tests/functional_composition/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -512,15 +476,12 @@ const resolve = createSelector(
-  [getIds, getObjects],
-  (ids, objects) => ids.map(id => objects[id])
+const resolve = createSelector([getIds, getObjects], (ids, objects) =>
+  ids.map(id => objects[id])
```
Previously looks better

brainkim replied:
🤷‍♂️ can’t win every battle. Lemme just mention that all the ramda examples are now formatted [exactly as they are in the documentation](https://ramdajs.com/docs/#compose) 🙂🙃🙂🙃

evilebottnawi commented:
Also CI failed, need fix it

brainkim commented:
@evilebottnawi my bad didn’t realize docs needed to be linted

lydell commented:
- I like the approach of looking for function literals instead of using that list of names.
- I think most of the changed tests look better.
- I have not reviewed the implementation code. (I've just glanced at those `for` loops and made a thinking face.)
- @brainkim I'm pretty confident that if we make a change here there will be follow-up issues, so be ready for making follow-up PRs after release :)

brainkim commented:
@lydell

> * I have not reviewed the implementation code. (I've just glanced at those for loops and made a thinking face.)

I have converted to using iterators (`for…of`) in my personal code but couldn’t find any usage of iterators in prettier so stuck with old-fashioned incrementing for loops.

> * @brainkim I'm pretty confident that if we make a change here there will be follow-up issues, so be ready for making follow-up PRs after release :)

Sure thing.

lipis commented:
> * I think most of the changed tests look better.

I second that!

j-f1 commented on src/language-js/printer-estree.js:
```diff
@@ -3863,40 +3863,34 @@ function isSimpleFlowType(node) {
+        return true;
+      }
+    } else if (arg.type === "CallExpression") {
+      for (let j = 0; j < arg.arguments.length; j++) {
```
Should this recurse?

```js
if (looksLikeFunctionCompositionArgs(arg.arguments)) {
  return true;
}
```

brainkim replied:
I initially thought I would have to write a recursive function as well, but after thinking about it, I decided against it because:
1. I was worried about the additional performance cost of walking the AST while walking the AST
2. checking at a depth of 1 is sufficient for all of the examples of function composition that I’ve found
3. in the course of walking the AST, any nested call expressions with function literals will split on multiple lines, which will automatically trigger the logic for splitting the parent call onto multiple lines.


If you can come up with counter-examples which would be uglier because we don’t check deeper in the tree for function literals I’ll be glad to change it, but honestly I can’t imagine a single example which wouldn’t trigger a line-split, if only because we’d blow the 80 character line budget.

j-f1 replied:
Thanks, that makes sense!

evilebottnawi commented:
/cc @brainkim want to clarify: which issues we can close after merge?

brainkim commented:
@evilebottnawi pretty much all of the open issues in the list in the first comment. Would you like me to explicitly reference these issues in the changelog or a commit?

#4751
#4858
#5060
#5769
#5928 <- this issue has an unrelated complaint related to object destructuring in arguments
#5969

evilebottnawi commented:
@brainkim just clarify what we can close after merge

brainkim commented:
Closes #4751 #4858 #5060 #5769 #5928 #5969

evilebottnawi commented:
/cc @prettier/core

duailibe commented:
@brainkim 

Thanks for this, sorry I took a while. I think we did more harm than good with that change. But we'll never get everything right.

Personally, as much as I don't like going back and forth, I think this one is worth it. This will certainly raise some issues from the people who asked for the "only identifiers as arguments" case.

I'll tag @vjeux and @suchipi who had some opinions on the original change. What do you think?

duailibe commented on src/language-js/printer-estree.js:
```diff
@@ -3979,7 +3972,7 @@ function printArgumentsList(path, options, print) {
   //    )
   // here, but not
   //    process.stdout.pipe(socket)
```
We might want to change this comment now

brainkim replied:
Moved and updated comment (let me know if you think no comment is necessary)

duailibe commented on src/language-js/printer-estree.js:
```diff
@@ -3863,40 +3863,33 @@ function isSimpleFlowType(node) {
+      arg.type === "FunctionExpression" ||
+      arg.type === "ArrowFunctionExpression"
+    ) {
+      if (++functionCount > 1) {
```
It's just a nit, but can you separate the increase from the `if` check? I personally like that, but we try to avoid those in codebase

brainkim replied:
My bad, I had it split up at first, bu noticed y’all had prefix increment operators in a couple places and changed it.

brainkim commented:
> This will certainly raise some issues from the people who asked for the "only identifiers as arguments" case.

Can you clarify what you mean by this case? I could take a look at it.

duailibe commented:
@brainkim 

```js
pipe(
  generate,
  mapToSomething,
  filter,
  reduce
);
```

I think it's reasonable to inline given the trade-offs, but it was requested and there's a chance people will complain it was reverted.

brainkim commented:
@duailibe

Yeah, I don’t think it’s necessary or “prettier” to deal with that case as explained in my comment [here](https://github.com/prettier/prettier/pull/6033#issuecomment-480093231). Isn’t one of the features of higher-order functions like `compose` and `flow` that you can write pretty one-liners? 🤷‍♂️

Almost all docs for higher-order functions put named functions on the same line. https://lodash.com/docs/#flow for instance

duailibe commented:
@brainkim We agree on this

brainkim commented:
@duailibe Feel free to ping if you need me to (destructively) rebase against the main branch I’m glad to help!

suchipi commented:
I'm anxious about changing this; last time we touched it, a lot of issues got opened where people who wanted the current formatting were asking why it got reverted. But I don't really have enough energy to defend the current implementation, so all I'm really saying is: be careful

brainkim commented:
@suchipi I’m fine with putting a hold on this until a major release. I’m not waiting for this PR to be merged for any reasons besides wanting to clear pending PRs in a timely fashion.

brainkim commented:
@duailibe Can we put a hold on this? I was looking over the issues for a second time and found an additional possible heuristic we can use from #1420, which is that if a call expression is immediately invoked (i.e. a curried function) we might consider breaking out the curried args onto multiple lines. Doing this will make connect look like this:

```js
const ArtistInput = connect(
  mapStateToProps,
  mapDispatchToProps,
  mergeProps
)(Component);
```

brainkim commented:
Here’s an initial pass at that logic: https://github.com/brainkim/prettier/commit/0d4c1a5a1667aa0b6d8d75e6dbe0179b4c82519f

The thing I’m struggling with is that I need to check if code would usually break in the second argument list and break in the first instead, otherwise, we’ll end up breaking a lot of tests, and I don’t yet know how to implement that because it’s logic involving multiple levels of the tree and finding out if the result before modification would normally cause a linebreak.

Lemme know if you think this should be done in a separate PR.

brainkim commented:
@duailibe @lydell @j-f1 @suchipi @evilebottnawi
I took a second pass at this pr last night when I couldn’t sleep. Sorry for letting it go stale just didn’t get around to working on this. I figured out a way to make curried calls break on the first group of arguments after reading [this comment about prioritizing line-breaks in sibling groups](https://github.com/prettier/prettier/issues/3014#issuecomment-376672416). The heuristic I use to detect long curried arguments is somewhat simple but it requires no changes to the test examples and covers most of the cases of ugly argument splitting described here: https://github.com/prettier/prettier/issues/1420.

brainkim commented:
Can someone review the new code though? I have literally no idea how the printer/builder commands work and I just sorta did guess and check until it printed what I wanted it to.

brainkim commented:
I added another bit of ungrouping logic for member expressions.

This reverts the breaking change where
```js
R.compose(
  Math.abs,
  R.add(1),
  R.multiply(2)
)(-4); //=> 7
```
gets printed as 

```js
R.compose(Math.abs, R.add(1), R.multiply(2))(-4) //=> 7
```

(https://github.com/prettier/prettier/pull/6033/commits/2e61e36ed2464b0903eae280d0eae666cb0c6073#diff-3bcf58dd9a3467170995691f1c4d741bL312). Now that I look at it, splitting curried calls across multiple lines feels like a good idea anyways because it’s not always obvious that you’re calling a call expression when it’s on a single line.

I also added another test case for the following example from #1420:
```
const mapStateToProps = state => ({
  users: R.compose(
    R.filter(R.propEq("status", "active")),
    R.values
  )(state.users)
});
```

At this point, I’m confident in saying almost all users won’t notice the change and we can finally get rid of using hard-coded strings (🤮🤮🤮🤮🤮)  to detect function composition.

lipis commented:
Is there anything else left in this PR?

brainkim commented:
@lipis not that I can tell. Maybe a review from someone who is familiar with the printer/builder code.

lipis commented:
There are bunch of approval.. I'll merge tomorrow if there are no objections.

brainkim commented:
Merge at your leisure! Getting it merged before the next minor release would be pretty chill though 😎😜👌

duailibe commented:
I was away on vacation for the last couple of weeks, sorry for the delay @brainkim

duailibe commented on src/language-js/printer-estree.js:
```diff
@@ -3976,40 +3984,55 @@ function isSimpleTemplateLiteral(node) {
+      functionCount++;
+      if (functionCount > 1) {
+        return true;
+      }
```
I believe you can skip the incremention and just return `true` directly :-)

And thus remove the `functionCount` entirely

brainkim replied:
The count is > 1 meaning we don’t trigger function composition rules when there’s only one function (e.g. `setTimeout(resolve, 100);`).

duailibe replied:
Oops my bad :-)

duailibe commented on src/language-js/printer-estree.js:
```diff
@@ -3976,40 +3984,55 @@ function isSimpleTemplateLiteral(node) {
+    return false;
+  }
+  let functionCount = 0;
+  for (let i = 0; i < args.length; i++) {
```
```suggestion
  for (const arg of args) {
```

duailibe commented:
I think it's a fair compromise, but I think at this point it's pretty obvious we won't please everyone.

evilebottnawi commented:
>  I think at this point it's pretty obvious we won't please everyone.
Agree, it is looks very good

lipis commented:
What's the status with this one?

evilebottnawi commented:
/cc @prettier/core friendly ping again

suchipi commented:
I still feel the same way as I did in https://github.com/prettier/prettier/pull/6033#issuecomment-492349702

That said, if you are fairly confident with the implementation, let's go for it

duailibe commented:
@lipis I'm aware of this PR, don't worry

brainkim commented:
I guess this is not being considered for 1.19? 😭😔😌

evilebottnawi commented:
/cc @duailibe

brainkim commented:
To be clear, I’m fine with this not being in the next release but maybe it could take merge priority over refactorings like https://github.com/prettier/prettier/pull/6562/files given the age of the PR. 🥺

evilebottnawi commented:
Added to `1.19` Milestone, need more view from prettier team

lydell commented:
I think we should be brave and merge this and see what feedback we get.

But.

The example in the changelog needs to better. It needs to show the worst output of the current behavior, and the best output of this PR. Currently, the example in the changelog is a little confusing, and kind of makes it look like this PR makes thing worse.

Also – have we explored looking at if arguments to a function are function calls? For example we could format like this:

```js
 .pipe(
    filter(isEven),
    map(double),
    scan(add, 0)
  )
```

Because the `.pipe` call has … at least 2 (3?) arguments and all (some?) are function calls? This might have unintended side effects though, but maybe our tests will call them out.

lipis commented:
> I think we should be brave and merge this

Let's do this!

evilebottnawi commented on tests/arrows/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -908,7 +908,10 @@ foo(c, a => b, d)
+promise.then(
+  result => result,
+  err => err
+);
```
I afraid it will be annoy many

lydell replied:
Hehe, this is one of the cases I think became better with this PR 😆

brainkim replied:
Typically, most `promise.then` calls are long enough that they’ll be split on multiple lines anyways.

brainkim commented:
@lydell
I replaced the example in the CHANGELOG to a starker, simpler example (from https://github.com/prettier/prettier/issues/5769).

```js
// Input
eventStore.update(id, _.flow(updater, incrementVersion))

// Output (Prettier stable)
eventStore.update(
  id,
  _.flow(
    updater,
    incrementVersion
  )
);

// Output (Prettier master)
eventStore.update(id, _.flow(updater, incrementVersion))
```

> Also – have we explored looking at if arguments to a function are function calls?

Wouldn’t this require the type information of identifiers? I was unaware of any such functionality in the prettier code base or rules which allowed us to do this. Additionally, as I argued elsewhere I think putting the arguments to pipe on the same line looks better when the arguments consist of identifiers and not function literals.

```js
  .pipe(filter(isEven), map(double), scan(add, 0))
```

lydell commented:
> Wouldn’t this require the type information of identifiers?

I meant CallExpression AST nodes.

brainkim commented:
> I meant CallExpression AST nodes.

I think the problem with that approach is that there would be too many false positives, where a statement like:

```js
multiply(add(1, 3), add(-2, -2));
```

would be split up on multiple lines. That would be a broader, potentially more disruptive change to prettier.

evilebottnawi commented:
Will be great if somebody run this on big code base, but this is just a wish.
Anyway looks good and i think we can shipped this in next version.

lydell commented on CHANGELOG.unreleased.md:
```diff
@@ -284,6 +284,30 @@ foo;
 
+### Javascript: Use function literals in arguments to detect function composition ([#6033] by [@brainkim])
+
+Previously, we used names of functions and methods to detect function
```
We should mention the history here. That we used a hardcoded list of "fp" function names, and that users of non-fp code witht he same name were confused why their code broke on multiple lines. And that we’re now trying out a new heuristic that we’re excited about, and that people are welcome to provide feedback, etc. Kind of.

brainkim replied:
I gave it a shot. I appreciate the idea that the CHANGELOG should be used to sell a specific feature and motivate users to upgrade. If you have further thoughts on the prose or language we should use lemme know.

brainkim commented:
Feel free to ping me for any issues caused by this PR. I promise to triage and attempt to fix any issues related to this change within a week for the next six months.

ThiefMaster commented:
This is screwing up the format of `createSelector` calls and making it much less readable. For example:

```diff
-export const isOngoingBooking = createSelector(
-  getDetailsWithRoom,
-  ({startDt, endDt}) => {
-    return moment().isBetween(startDt, endDt, 'day');
-  }
-);
+export const isOngoingBooking = createSelector(getDetailsWithRoom, ({startDt, endDt}) => {
+  return moment().isBetween(startDt, endDt, 'day');
+});
```
